vms/platformvm/state/diff.go

```diff
@@ -76,6 +81,201 @@ func NewDiff(
+	itAddedCurrentStakers := NewTreeIterator(d.currentStakerDiffs.addedStakers)
+	defer itAddedCurrentStakers.Release()
+	for itAddedCurrentStakers.Next() {
+		var (
+			staker     = itAddedCurrentStakers.Value()
+			stakerTxID = staker.TxID
+		)
+		tx, found := d.addedTxs[stakerTxID]
+		if !found {
+			return nil, fmt.Errorf("added staker %v, but can't find its tx among added ones", stakerTxID)
+		}
+
+		key := merkleCurrentStakersKey(stakerTxID)
+		stakerData := stakersData{
+			TxBytes:         tx.tx.Bytes(),
+			PotentialReward: staker.PotentialReward,
+		}
+		dataBytes, err := txs.GenesisCodec.Marshal(txs.Version, stakerData)
+		if err != nil {
+			return nil, fmt.Errorf("failed to serialize current stakers data, stakerTxID %v: %w", stakerTxID, err)
+		}
+		batchOps = append(batchOps, database.BatchOp{
+			Key:   key,
+			Value: dataBytes,
+		})
+	}
+	for txID := range d.currentStakerDiffs.deletedStakers {
+		key := merkleCurrentStakersKey(txID)
+		batchOps = append(batchOps, database.BatchOp{
+			Key:    key,
+			Delete: true,
+		})
+	}
+
+	// PENDING STAKERS
+	itAddedPendingStakers := NewTreeIterator(d.pendingStakerDiffs.addedStakers)
+	defer itAddedPendingStakers.Release()
+	for itAddedPendingStakers.Next() {
+		var (
+			staker     = itAddedPendingStakers.Value()
+			stakerTxID = staker.TxID
+		)
+		tx, found := d.addedTxs[stakerTxID]
+		if !found {
+			return nil, fmt.Errorf("added staker %v, but can't find its tx among added ones", stakerTxID)
+		}
+
+		key := merkleCurrentStakersKey(stakerTxID)
+		stakerData := stakersData{
+			TxBytes:         tx.tx.Bytes(),
+			PotentialReward: staker.PotentialReward,
+		}
+		dataBytes, err := txs.GenesisCodec.Marshal(txs.Version, stakerData)
+		if err != nil {
+			return nil, fmt.Errorf("failed to serialize current stakers data, stakerTxID %v: %w", stakerTxID, err)
+		}
+		batchOps = append(batchOps, database.BatchOp{
+			Key:   key,
+			Value: dataBytes,
+		})
+	}
+	for txID := range d.pendingStakerDiffs.deletedStakers {
+		key := merkleCurrentStakersKey(txID)
+		batchOps = append(batchOps, database.BatchOp{
+			Key:    key,
+			Delete: true,
+		})
+	}
+
+	// DELEGATEE REWARDS
+	for subnetID, nodeDelegateeRewards := range d.modifiedDelegateeRewards {
+		for nodeID, delegateeReward := range nodeDelegateeRewards {
+			key := merkleDelegateeRewardsKey(nodeID, subnetID)
+			batchOps = append(batchOps, database.BatchOp{
+				Key:   key,
+				Value: database.PackUInt64(delegateeReward),
+			})
+		}
+	}
+
+	// UTXOS
+	for utxoID, utxo := range d.modifiedUTXOs {
+		key := merkleUtxoIDKey(utxoID)
+		if utxo == nil { // delete the UTXO
+			batchOps = append(batchOps, database.BatchOp{
+				Key:    key,
+				Delete: true,
+			})
+			continue
+		}
+
+		utxoBytes, err := txs.GenesisCodec.Marshal(txs.Version, utxo)
+		if err != nil {
+			return nil, fmt.Errorf("failed marshalling utxo %v bytes: %w", utxoID, err)
+		}
+		batchOps = append(batchOps, database.BatchOp{
+			Key:   key,
+			Value: utxoBytes,
+		})
+	}
```

Comment:
this is basically duplicated code from state.State.
If we think this is the correct way to proceed we can refactor it. move serialization of key/value pairs when each item is inserted, or maybe even just store all this stuff as key/pair value in merkleDB and get rid of all the diff and state caches shit